[Alerting]: Fix updating rule group and add tests

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
* Fix updating rule labels
* Set default values for rule no data and error states
if they are missing
* Add test for updating rule
* Add tests for posting invalid alert rules
* Fix alert rule validation for missing queries and expressions

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

papagian committed:
[Alerting]: Fix updating rule group and add test

* Fix updating rule labels
* Set default values for rule no data and error states
if they are missing
* Add test for updating rule

papagian committed:
small fixes

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-fix-update-labels

papagian committed:
fix after syncing with master

papagian committed:
Test updating annotations

papagian committed:
Remove print statements

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -73,12 +75,23 @@ func TestAlertRuleCRUD(t *testing.T) {
+						Annotations: map[string]string{"annotation1": "val1"},
+					},
+					// this rule does not explicitly set no data and error states
+					// thererfore it should get the default values
```
```suggestion
					// therefore it should get the default values
```

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -138,7 +153,6 @@ func TestAlertRuleCRUD(t *testing.T) {
 		b, err := ioutil.ReadAll(resp.Body)
 		require.NoError(t, err)
 
-		fmt.Println(string(b))
```
🤦 thanks!

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -73,12 +75,23 @@ func TestAlertRuleCRUD(t *testing.T) {
 			Rules: []apimodels.PostableExtendedRuleNode{
 				{
+					ApiRuleNode: &apimodels.ApiRuleNode{
+						For:         dur,
```
```suggestion
						For:         interval,
```

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -73,12 +75,23 @@ func TestAlertRuleCRUD(t *testing.T) {
 	// Create the namespace we'll save our alerts to.
 	require.NoError(t, createFolder(t, store, 0, "default"))
 
+	dur, err := model.ParseDuration("1m")
```
```suggestion
	interval, err := model.ParseDuration("1m")
```

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -230,24 +251,71 @@ func TestAlertRuleCRUD(t *testing.T) {
       }
    ]
-}`, rulesNamespaceWithoutVariableValues(t, b))
+}`, rulesNamespaceWithoutVariableValues(t, b, ruleUIDs))
```
I think it's better semantically if make this function return the UIDs rather than passing the argument

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -230,24 +251,71 @@ func TestAlertRuleCRUD(t *testing.T) {
-		// If the rule group name does not exists
-		u := fmt.Sprintf("http://%s/api/ruler/grafana/api/v1/rules/default/groupnotexist", grafanaListedAddr)
-		req, err := http.NewRequest(http.MethodDelete, u, nil)
+		dur, err := model.ParseDuration("30s")
```
```suggestion
		interval, err := model.ParseDuration("30s")
```

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -230,24 +251,71 @@ func TestAlertRuleCRUD(t *testing.T) {
+					},
+					GrafanaManagedAlert: &apimodels.PostableGrafanaRule{
+						OrgID:     2,
+						UID:       ruleUIDs[0],
```
```suggestion
						UID:       ruleUIDs[0], // Including the UID in the payload makes the endpoint update the existing rule.
```

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -230,24 +251,71 @@ func TestAlertRuleCRUD(t *testing.T) {
 
-	client := &http.Client{}
-	// Finally, make sure we can delete it.
+	// update the first rule and completely remove the other
```
Should we add a case to try and update a rule that does not exist?

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -230,24 +251,71 @@ func TestAlertRuleCRUD(t *testing.T) {
 
-	client := &http.Client{}
-	// Finally, make sure we can delete it.
+	// update the first rule and completely remove the other
```
Are we? This reads as if we're updating one rule and deleting another, but we're actually updating one rule and then deleting the whole rule group.

papagian replied:
I mean that the result of this POST request will be to update the 1st rule created in [before](https://github.com/grafana/grafana/blob/3ae4df2cfe1dcbdd1f87daea666e7d66339007da/pkg/tests/api/alerting/api_alertmanager_test.go#L147) and delete the 2nd one because is not included in the payload.

gotjosh replied:
Makes sense, I'd include that as part of the description to make it clear. I was surprised given we include the ID of the rule in the payload... so I thought it would only update what we included an ID for.

papagian committed:
Apply suggestions from code review

Co-authored-by: gotjosh <josue@grafana.com>

papagian committed:
Improve tests following review comments

* add test for posting an unknown rule UID
* modify rulesNamespaceWithoutVariableValues to return the rule UIDs
instead of get them as parameter

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=33074) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=33074) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=33074) it.</sub>

papagian committed:
Fix alert rule validation and add tests

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-fix-update-labels

papagian committed:
Remove org id from PostableGrafanaRule

This field was not used; each rule gets the organisation of the user making
the rerquest

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -296,14 +557,22 @@ func createFolder(t *testing.T, store *sqlstore.SQLStore, folderID int64, folder
 }
 
 // rulesNamespaceWithoutVariableValues takes a apimodels.NamespaceConfigResponse JSON-based input and makes the dynamic fields static e.g. uid, dates, etc.
```
```suggestion
// rulesNamespaceWithoutVariableValues takes a apimodels.NamespaceConfigResponse JSON-based input and makes the dynamic fields static e.g. uid, dates, etc.
// it returns a map of the modified rule UIDs with the namespace,rule_group as a key
```

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-fix-update-labels

papagian committed:
Update pkg/tests/api/alerting/api_alertmanager_test.go

Co-authored-by: gotjosh <josue@grafana.com>

papagian committed:
fixup

papagian committed:
fixup

papagian committed:
Fix legacy test
